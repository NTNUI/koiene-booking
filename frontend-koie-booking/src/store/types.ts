// rootState = generic type, defines the root state type
import AdminBooking, { AdminBookingDictionary } from '@/types/admin/AdminBooking';
import { AdminReportKoieDictionary } from '@/types/admin/AdminReports';

export interface RootState {
  version: string;
  koie: KoieState;
  booking: BookingState;
  auth: AuthState;
  avalanche: AvalancheState;
  adminBookings: AdminBookingsState;
}

export interface KoieState {
  currentLocale: string;
  isLoading: boolean;
  koieData: Koie;
  error: boolean;
  allKoier: [];
}

export interface Koie {
  location: {
    latitude: string;
    longitude: string;
    area: string;
    difficultyInfo_eng: string;
    difficultyInfo_nor: string;
    altitude: number;
    terrain_eng: string;
    terrain_nor: string;
    map_pdf: string;
    kartblad: string;
  };
  description: {
    yr_link_eng: string;
    yr_link_nor: string;
    description_eng: string;
    description_nor: string;
    directions_eng: string;
    directions_nor: string;
    parking_eng: string;
    parking_nor: string;
  };
  beds_available_in_booking_window: Array<BedsAvailable>;
  name: string;
  number_of_beds: number;
  booking_window: number;
  price_member: number;
  price_not_member: number;
}

export interface BedsAvailable {
  date: string;
  bedsAvailable: number;
}

export interface BookingState {
  isLoading: boolean;
  dateFrom: string;
  dateTo: string;
  beds: number;
  step: number;
  guests: Array<Guest>;
  numberOfMembers: number;
  numberOfNonMembers: number;
  edited: boolean;
  validForm: boolean;
  unavailableDates: [];
  availableBeds: number;
  noDatesAvailable: boolean;
  error: boolean;
  bookingData: BookingInfo;
  paymentConfirmed: boolean;
}

export interface ReportState {
  validForm: boolean;
  edited: boolean;
  isLoading: boolean;
  error: boolean;
  step: number;
  reportData: ReportData;
  bookingInfo: BookingInfo;
}

export interface ReportData {
  id: number;
  booking_uuid: string;
  firewood: number;
  chopped_up_wood: number;
  smoke_detector_is_working: boolean;
  gas_is_full: boolean;
  gas_burner_primus: number;
  axe: number;
  hammer: number;
  saw: number;
  saw_blade: number;
  saw_bench: number;
  spade: number;
  kerosene_lamp: number;
  detergent: number;
  dishware: number;
  cookware: number;
  cabin_book: number;
  candle_holders: number;
  fire_blanket: number;
  fire_extinguisher: number;
  other_faults: string;
  boat_status: number;
  canoe_status: number;
  life_jackets_status: number;
  feedback: string;
}

export interface AvalancheState {
  warningLevels: Array<Warning>;
  disableBooking: boolean;
  warningData: Array<WarningResult>;
  koierToCheck: Array<string>;
}

export interface Warning {
  day: string;
  level: number;
}

export interface WarningResult {
  dangerLevel: number;
  validTo: string;
}

export interface CreateBookingInfo {
  koie: string;
  from_date: string;
  to_date: string;
  guests: Array<Guest>;
  guests_member: number;
  guests_not_member: number;
}

export interface BookingInfo {
  id: number;
  uuid: string;
  user: number;
  koie: string;
  price: number;
  booking_transaction_id: string;
  from_date: string;
  to_date: string;
  guests_member: number;
  guests_not_member: number;
}

export interface AuthState {
  isAdmin: boolean;
  tokens: {
    access: string | null;
    refresh: string | null;
  };
}

export interface Guest {
  name: string;
  number: string;
  email: string;
  isMember: boolean;
}

export interface AdminBookingsState {
  cabinsWithBookings: AdminBookingDictionary;
  startDate: string;
}

export interface AdminReportsState {
  chosenCabin: string;
  expanded: boolean;
  cabins: AdminReportKoieDictionary;
}
