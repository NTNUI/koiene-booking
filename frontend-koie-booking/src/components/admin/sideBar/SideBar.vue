<template>
  <v-card>
    <v-navigation-drawer id="sideBarDrawer" v-model="drawer" :color="$scssVars.globalColorPrimary" permanent dark app>
      <v-list>
        <SideBarTop />
        <SideBarBottom />
      </v-list>
    </v-navigation-drawer>
  </v-card>
</template>

<script lang="ts">
import Vue from 'vue';
import SideBarTop from '@/components/admin/sideBar/SideBarTop.vue';
import SideBarBottom from '@/components/admin/sideBar/SideBarBottom.vue';

export default Vue.extend({
  name: 'SideBar',
  components: { SideBarBottom, SideBarTop },
  props: {
    currentView: {
      type: Number,
      default: 0
    }
  },
  data() {
    return {
      drawer: this.currentView
    };
  },
  watch: {
    currentView(newVal: number) {
      this.drawer = newVal;
    }
  }
});
</script>

<style lang="scss">
#sideBarDrawer {
  border-top-left-radius: 0 !important;
  border-top-right-radius: 0 !important;
}
</style>
