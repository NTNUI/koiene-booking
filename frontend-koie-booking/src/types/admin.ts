import { Component } from 'vue';

export interface AdminView {
  id: String; // Used to reference the buttons in SideBar
  title: String;
  icon: String;
  component: Component;
}
