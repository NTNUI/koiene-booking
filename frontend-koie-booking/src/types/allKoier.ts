export interface AllKoierData {
  koieTitle: string;
}
