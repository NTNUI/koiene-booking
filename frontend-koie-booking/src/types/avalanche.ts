import { TranslateResult } from 'vue-i18n';

export interface AvalancheData {
  result: string;
  levelFirstDay: string;
  levelSecondDay: string;
  levelThirdDay: string;
  firstDay: string;
  secondDay: string;
  thirdDay: string;
  fromDate: string;
  nextDate: string;
  lastDate: string;
  warningTypes: Array<TranslateResult>;
  colors: Array<string>;
}
