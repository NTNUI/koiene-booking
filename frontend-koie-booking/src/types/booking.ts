import { TranslateResult } from 'vue-i18n';

export interface BookingData {
  step: number;
  labels: Array<TranslateResult>;
}

export interface BookingFirstStepData {
  bedsTotal: number;
  bedsChosen: number;
  error: boolean;
}

export interface BookingSecondStepData {
  bedsChosen: number;
  bookingStep: number;
  guests: Array<Guest>;
  numberRules: object;
  nameRules: object;
  emailStrictRules: object;
  emailLooseRules: object;
  edited: boolean;
  validForm: boolean;
  memberPrice: number;
  nonMemberPrice: number;
}

export interface BookingThirdStepData {
  complete: boolean;
  stripeOptions: object;
  totalPrice: number;
}

export interface Guest {
  name: string;
  number: string;
  isMember: boolean;
  email?: string;
}
