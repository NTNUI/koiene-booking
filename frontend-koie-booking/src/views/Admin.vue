<template>
  <div>
    <SideBar :current-view="currentView" />
    <AdminContentView :admin-component="adminViews[currentView].component" />
  </div>
</template>

<script lang="ts">
import Vue from 'vue';

import AdminContentView from '@/components/admin/AdminContentView.vue';
import SideBar from '@/components/admin/sideBar/SideBar.vue';
import adminViews from '@/components/admin/AdminViews';

export default Vue.extend({
  name: 'Admin',
  components: { AdminContentView, SideBar },
  provide() {
    return {
      getCurrentView: () => this.currentView,
      setCurrentView: (index: number) => {
        this.updateView(index);
      },
    };
  },
  data() {
    return {
      adminViews: adminViews,
      currentView: 0,
    };
  },
  computed: {
    isAdmin(): boolean {
      return this.$store.getters['auth/isAdmin'];
    },
  },
  watch: {
    isAdmin: {
      handler: function(newVal: boolean) {
        if (!newVal) this.$router.push('/');
      },
      immediate: true,
    },
  },
  methods: {
    updateView(index: number): void {
      this.currentView = index;
    },
  },
});
</script>

<style scoped></style>
