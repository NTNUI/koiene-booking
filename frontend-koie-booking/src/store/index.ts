import Vue from 'vue';
import Vuex, { StoreOptions } from 'vuex';
import axios from 'axios';
import VueAxios from 'vue-axios';
import { RootState } from './types';
import { koie } from './modules/users/koie/index';
import { booking } from './modules/users/booking/index';
import { avalanche } from './modules/users/avalanche/index';
import { report } from './modules/users/report/index';
import { auth } from './modules/users/auth/index';
import { adminBookings } from '@/store/modules/admin/bookings';

Vue.use(Vuex, VueAxios, axios);

const store: StoreOptions<RootState> = {
  modules: {
    adminBookings,
    koie,
    booking,
    avalanche,
    report,
    auth
  }
};

export default new Vuex.Store<RootState>(store);

export { store as storeConfig };
