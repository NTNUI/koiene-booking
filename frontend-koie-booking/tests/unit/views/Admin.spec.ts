import Vue, { VueConstructor } from 'vue';
import Vuetify, { Vuetify as VuetifyType } from 'vuetify';
import Vuex, { Store } from 'vuex';
import { storeConfig } from '@/store';
import { cloneDeep } from 'lodash';
import scssVars from '@/styles/variables.scss';
import mockAxios from 'jest-mock-axios';

Vue.use(Vuetify);

// Utilities
import { mount, createLocalVue, Wrapper } from '@vue/test-utils';

// Components or views
import Admin from '@/views/Admin.vue';
import { RootState } from '@/store/types';
import SideBar from '@/components/admin/sideBar/SideBar.vue';
import adminViews from '@/components/admin/AdminViews';

describe('View Admin.vue', () => {
  let wrapper: Wrapper<any>;
  let localVue: VueConstructor<Vue>;
  let vuetify: VuetifyType;
  let store: Store<RootState>;

  beforeEach(() => {
    localVue = createLocalVue();
    vuetify = new Vuetify();
    localVue.use(Vuetify);
    localVue.use(Vuex);
    store = new Vuex.Store(cloneDeep(storeConfig));
    localVue.prototype.$scssVars = scssVars;

    wrapper = mount(Admin, {
      sync: false,
      localVue,
      vuetify,
      store,
      computed: {
        isAdmin() {
          return true;
        },
      },
    });
  });

  afterEach(() => {
    mockAxios.reset();
  });

  it('Matches snapshot', () => {
    // Assert
    expect(wrapper).toMatchSnapshot();
  });

  it('Renders sidebar', () => {
    // Assert
    expect(wrapper.contains(SideBar)).toBe(true);
  });

  it('Admin view renders correct component view', () => {
    // Assert
    expect(wrapper.contains(adminViews[0].component)).toBe(true);
  });

  it("Admin view doesn't render wrong view", () => {
    // Assert
    expect(wrapper.contains(adminViews[1].component)).toBe(false);
  });

  it('Admin view updates rendered component view', async () => {
    for (let i = adminViews.length - 1; i > -1; i--) {
      // Act
      await wrapper.find('#sideBar' + adminViews[i].id).trigger('click');

      // Assert
      expect(wrapper.contains(adminViews[i].component)).toBe(true);
    }
  });
});
