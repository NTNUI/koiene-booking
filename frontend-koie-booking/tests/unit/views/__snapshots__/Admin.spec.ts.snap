// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Admin.vue Matches snapshot 1`] = `
<div>
  <div class="v-card v-sheet theme--light">
    <nav class="v-navigation-drawer v-navigation-drawer--fixed v-navigation-drawer--open theme--dark" style="height: 100vh; top: 0px; max-height: calc(100% - 0px); transform: translateX(0%); width: 256px;" id="sideBarDrawer">
      <div class="v-navigation-drawer__content">
        <div class="v-list v-sheet v-sheet--tile theme--dark">
          <div tabindex="-1" class="v-list-item v-list-item--two-line theme--dark">
            <div class="v-avatar v-list-item__avatar" style="height: 40px; min-width: 40px; width: 40px;"><img src="https://ntnui.no/wp-content/uploads/2018/04/ntnui-logo-liten.png"></div>
            <div class="v-list-item__content">
              <div class="v-list-item__title">Admin</div>
              <div class="v-list-item__subtitle">Koie-styret</div>
            </div>
          </div>
          <div>
            <div tabindex="0" id="sideBarallCabins" class="v-list-item v-list-item--link theme--dark">
              <div class="v-list-item__icon"><i aria-hidden="true" v-text="calendar-month" class="v-icon notranslate material-icons theme--dark"></i></div>
              <div class="v-list-item__content">
                <div class="v-list-item__title">Alle koier</div>
              </div>
            </div>
            <div tabindex="0" id="sideBarreports" class="v-list-item v-list-item--link theme--dark">
              <div class="v-list-item__icon"><i aria-hidden="true" v-text="file-document-outline" class="v-icon notranslate material-icons theme--dark"></i></div>
              <div class="v-list-item__content">
                <div class="v-list-item__title">Rapporter</div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="v-navigation-drawer__border"></div>
    </nav>
  </div>
  <div id="adminContentView">
    <div class="row justify-center">
      <h4>Her kommer tabell over alle hytter</h4>
    </div>
  </div>
</div>
`;
